src/lib/pickles/dlog_main_inputs.ml

```diff
@@ -0,0 +1,195 @@
+    module Impl = Impl
+
+    module Params = struct
+      include G1.Params
+
+      let one = G1.to_affine_exn G1.one
+
+      let group_size_in_bits = Field.size_in_bits
+    end
+
+    module F = struct
+      include struct
+        open Impl.Field
+
+        type nonrec t = t
+
+        let ( * ), ( + ), ( - ), inv_exn, square, scale, if_, typ, constant =
+          (( * ), ( + ), ( - ), inv, square, scale, if_, typ, constant)
+
+        let negate x = scale x Constant.(negate one)
+      end
+
+      module Constant = struct
+        open Impl.Field.Constant
+
+        type nonrec t = t
+
+        let ( * ), ( + ), ( - ), inv_exn, square, negate =
+          (( * ), ( + ), ( - ), inv, square, negate)
+      end
+
+      let assert_square x y = Impl.assert_square x y
+
+      let assert_r1cs x y z = Impl.assert_r1cs x y z
+    end
+
+    module Constant = struct
+      include G1.Affine
+      module Scalar = Impls.Pairing_based.Field.Constant
+
+      let scale (t : t) x : t = G1.(to_affine_exn (scale (of_affine t) x))
+
+      let random () = G1.(to_affine_exn (random ()))
+
+      let negate x = G1.(to_affine_exn (negate (of_affine x)))
+
+      let zero = Impl.Field.Constant.(zero, zero)
+
+      let ( + ) t1 t2 =
+        let is_zero (x, _) = Impl.Field.Constant.(equal zero x) in
+        if is_zero t1 then t2
+        else if is_zero t2 then t1
+        else
+          let r = G1.(of_affine t1 + of_affine t2) in
+          try G1.to_affine_exn r with _ -> zero
+
+      let to_affine_exn = Fn.id
+
+      let of_affine = Fn.id
+    end
+  end
+
+  module Params = Inputs.Params
+  module Constant = Inputs.Constant
+  module T = Snarky_curve.For_native_base_field (Inputs)
+
+  include (
+    T :
+      module type of T
+      with module Scaling_precomputation := T.Scaling_precomputation )
+
+  module Scaling_precomputation = T.Scaling_precomputation
+
+  let ( + ) = add_exn
+
+  let scale t bs = T.scale t (Bitstring_lib.Bitstring.Lsb_first.of_list bs)
+
+  let to_field_elements (x, y) = [x; y]
+
+  let assert_equal (x1, y1) (x2, y2) =
+    Field.Assert.equal x1 x2 ; Field.Assert.equal y1 y2
+
+  let scale_inv t bs =
+    let res =
+      exists typ
+        ~compute:
+          As_prover.(
+            fun () ->
+              G1.(
+                to_affine_exn
+                  (scale
+                     (of_affine (read typ t))
+                     (Fp.inv (Fp.of_bits (List.map ~f:(read Boolean.typ) bs))))))
+    in
+    assert_equal t (scale res bs) ;
+    res
+
+  (* g -> 7 * g *)
+  let scale_by_quadratic_nonresidue t =
+    let t2 = T.double t in
+    let t4 = T.double t2 in
+    t + t2 + t4
+
+  let one_seventh = Fp.(inv (of_int 7))
+
+  let scale_by_quadratic_nonresidue_inv t =
+    let res =
+      exists typ
+        ~compute:
+          As_prover.(
+            fun () ->
+              G1.(to_affine_exn (scale (of_affine (read typ t)) one_seventh)))
+    in
+    ignore (scale_by_quadratic_nonresidue res) ;
```

Comment:
Shouldn't this be `assert_equal (scale_by_quadratic_nonresidue res) t`? At the moment, `res` seems to be unconstrained.

Reply:
ach -- yes, thanks not sure how it slipped through. will correct.